policies/recipes/steps.go

```diff
@@ -90,8 +100,335 @@ func parsePermissions(s string) (os.FileMode, error) {
+		decompressed.Close()
+		if err != nil {
+			return err
+		}
+		return extractTar(decompressed.Name(), step.ArchiveExtraction.Destination)
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_LZMA:
+		compressed, err := os.Open(filename)
+		if err != nil {
+			return err
+		}
+		defer compressed.Close()
+		reader, err := lzma.NewReader2(compressed)
+		if err != nil {
+			return err
+		}
+		decompressed, err := ioutil.TempFile(stepDir, "archive-*.tar")
+		if err != nil {
+			return err
+		}
+		_, err = io.Copy(decompressed, reader)
+		decompressed.Close()
+		if err != nil {
+			return err
+		}
+		return extractTar(decompressed.Name(), step.ArchiveExtraction.Destination)
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_XZ:
+		compressed, err := os.Open(filename)
+		if err != nil {
+			return err
+		}
+		defer compressed.Close()
+		reader, err := xz.NewReader(compressed)
+		if err != nil {
+			return err
+		}
+		decompressed, err := ioutil.TempFile(stepDir, "archive-*.tar")
+		if err != nil {
+			return err
+		}
+		_, err = io.Copy(decompressed, reader)
+		decompressed.Close()
+		if err != nil {
+			return err
+		}
+		return extractTar(decompressed.Name(), step.ArchiveExtraction.Destination)
+	default:
+		return fmt.Errorf("Unrecognized archive type %q", step.ArchiveExtraction.GetType())
+	}
+}
+
+func zipIsDir(name string) bool {
+	return strings.HasSuffix(name, string(os.PathSeparator))
+}
+
+func extractZip(zipPath string, dst string) error {
+	zr, err := zip.OpenReader(zipPath)
+	if err != nil {
+		return err
+	}
+	defer zr.Close()
+
+	// Check for conflicts
+	for _, f := range zr.File {
+		filen, err := common.NormPath(filepath.Join(dst, f.FileHeader.Name))
+		if err != nil {
+			return err
+		}
+		fmt.Printf("checking if file %s exists\n", filen)
+		stat, err := os.Stat(filen)
+		if os.IsNotExist(err) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if zipIsDir(filen) && stat.IsDir() {
+			// it's ok if directories already exist
+			continue
+		}
+		return fmt.Errorf("file exists: %s", filen)
+	}
+
+	// Create dirs
+	for _, f := range zr.File {
+		filen, err := common.NormPath(filepath.Join(dst, f.FileHeader.Name))
+		if err != nil {
+			return err
+		}
+
+		if !zipIsDir(filen) {
+			continue
+		}
+		_, err = os.Stat(filen)
+		if err == nil {
+			continue
+		}
+		if !os.IsNotExist(err) {
+			return err
+		}
+		err = os.MkdirAll(filen, 0755)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create files.
+	for _, f := range zr.File {
+		filen, err := common.NormPath(filepath.Join(dst, f.Name))
+		if err != nil {
+			return err
+		}
+		if zipIsDir(filen) {
+			continue
+		}
+		filedir := filepath.Dir(filen)
+		_, err = os.Stat(filedir)
+		if os.IsNotExist(err) {
+			err = os.MkdirAll(filedir, 0755)
+		}
+		if err != nil {
+			return err
+		}
+		fmt.Printf("os.Create %s\n", filen)
+		reader, err := f.Open()
+		if err != nil {
+			return err
+		}
+
+		dst, err := os.OpenFile(filen, os.O_RDWR|os.O_CREATE, 0755)
```

Comment:
why always 0755? we don't want to respect the file modes from the zip header?

Reply:
I'm under the impression that permissions are not a default part of the zip header. If it's an important feature I could look into how most zip utilities store them.

Reply:
i'm basing this entirely on just reading the go docs for the zip library. if it's populated i would think we'd want to respect it, but idk if it's usually populated.

Reply:
After looking at the docs in more detail you are correct, I was focused on the struct fields I didn't look too much at the methods apparently. I'll add the support for it in.